src/com/facebook/buck/go/GoBinary.java

```diff
@@ -0,0 +1,142 @@
+ * under the License.
+ */
+
+package com.facebook.buck.go;
+
+import com.facebook.buck.cxx.CxxPlatform;
+import com.facebook.buck.cxx.Linker;
+import com.facebook.buck.graph.AbstractBreadthFirstTraversal;
+import com.facebook.buck.model.BuildTargets;
+import com.facebook.buck.rules.AbstractBuildRule;
+import com.facebook.buck.rules.AddToRuleKey;
+import com.facebook.buck.rules.BuildContext;
+import com.facebook.buck.rules.BuildRule;
+import com.facebook.buck.rules.BuildTargetSourcePath;
+import com.facebook.buck.rules.BuildableContext;
+import com.facebook.buck.rules.CommandTool;
+import com.facebook.buck.rules.Tool;
+import com.facebook.buck.rules.BinaryBuildRule;
+import com.facebook.buck.rules.BuildRuleParams;
+import com.facebook.buck.rules.BuildableProperties;
+import com.facebook.buck.rules.SourcePathResolver;
+import com.facebook.buck.step.Step;
+import com.facebook.buck.step.fs.MakeCleanDirectoryStep;
+import com.facebook.buck.step.fs.MkdirStep;
+import com.facebook.buck.step.fs.SymlinkTreeStep;
+import com.google.common.base.Function;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.FluentIterable;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.io.Files;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+public class GoBinary extends AbstractBuildRule implements BinaryBuildRule {
+
+  @AddToRuleKey
+  private final Tool linker;
+  @AddToRuleKey
+  private final ImmutableList<String> linkerFlags;
+  @AddToRuleKey
+  private final Linker cxxLinker;
+
+  private final GoLinkable lib;
+
+  private final Path scratchDir;
+  private final Path output;
+
+  public GoBinary(
+      BuildRuleParams params,
+      SourcePathResolver resolver,
+      Linker cxxLinker,
+      GoLinkable lib,
+      Tool linker,
+      ImmutableList<String> linkerFlags) {
+    super(params, resolver);
+    this.cxxLinker = cxxLinker;
+    this.scratchDir = BuildTargets.getScratchPath(getBuildTarget(), "__%s__link");
+    this.linker = linker;
+    this.lib = lib;
+    this.output = BuildTargets.getGenPath(
+        params.getBuildTarget(), "%s/" + params.getBuildTarget().getShortName());
+    this.linkerFlags = linkerFlags;
+  }
+
+  @Override
+  public Tool getExecutableCommand() {
+    return new CommandTool.Builder()
+        .addArg(new BuildTargetSourcePath(getBuildTarget()))
+        .build();
+  }
+
+  @Override
+  public BuildableProperties getProperties() {
+    return new BuildableProperties(BuildableProperties.Kind.PACKAGING);
+  }
+
+  @Override
+  public ImmutableList<Step> getBuildSteps(
+      BuildContext context, BuildableContext buildableContext) {
+    final ImmutableMap.Builder<Path, Path> symlinkTree = ImmutableMap.builder();
+    AbstractBreadthFirstTraversal<GoLinkable> visitor = new AbstractBreadthFirstTraversal<GoLinkable>(lib) {
```

Comment:
We usually do this kind of visiting logic in the description, then pass the results of the visiting to the `BuildRule`'s constructor.